lib/babe/errors.go

```diff
@@ -81,88 +80,220 @@ func (e TransactionValidityError) Error() string {
+
+// Index Returns VDT index
+func (err BadMandatory) Index() uint { return 8 }
+
+// MandatoryDispatch A transaction with a mandatory dispatch
+type MandatoryDispatch struct{}
+
+// Index Returns VDT index
+func (err MandatoryDispatch) Index() uint { return 9 }
+
+func determineErrType(vdt scale.VaryingDataType) error {
+	switch val := vdt.Value().(type) {
+	case Other:
+		return &DispatchOutcomeError{fmt.Sprintf("unknown error: %s", val)}
+	case CannotLookup:
+		return &DispatchOutcomeError{"failed lookup"}
+	case BadOrigin:
+		return &DispatchOutcomeError{"bad origin"}
+	case Module:
+		return &DispatchOutcomeError{fmt.Sprintf("custom module error: %s", val.string())}
+	case Call:
 		return &TransactionValidityError{"call of the transaction is not expected"}
-	case 1:
+	case Payment:
 		return &TransactionValidityError{"invalid payment"}
-	case 2:
+	case Future:
 		return &TransactionValidityError{"invalid transaction"}
-	case 3:
+	case Stale:
 		return &TransactionValidityError{"outdated transaction"}
-	case 4:
+	case BadProof:
 		return &TransactionValidityError{"bad proof"}
-	case 5:
+	case AncientBirthBlock:
 		return &TransactionValidityError{"ancient birth block"}
-	case 6:
+	case ExhaustsResources:
 		return &TransactionValidityError{"exhausts resources"}
-	case 7:
-		return &TransactionValidityError{fmt.Sprintf("unknown error: %d", res[1])}
-	case 8:
+	case InvalidCustom:
+		return &TransactionValidityError{fmt.Sprintf("unknown error: %d", val)}
+	case BadMandatory:
 		return &TransactionValidityError{"mandatory dispatch error"}
-	case 9:
+	case MandatoryDispatch:
 		return &TransactionValidityError{"invalid mandatory dispatch"}
-	}
-	return errInvalidResult
-}
-
-func determineUnknownTxnErr(res []byte) error {
-	switch res[0] {
-	case 0:
+	case ValidityCannotLookup:
 		return &TransactionValidityError{"lookup failed"}
-	case 1:
+	case NoUnsignedValidator:
 		return &TransactionValidityError{"validator not found"}
-	case 2:
-		return &TransactionValidityError{fmt.Sprintf("unknown error: %d", res[1])}
+	case UnknownCustom:
+		return &TransactionValidityError{fmt.Sprintf("unknown error: %d", val)}
 	}
+
 	return errInvalidResult
 }
 
 func determineErr(res []byte) error {
-	switch res[0] {
-	case 0: // DispatchOutcome
-		switch res[1] {
-		case 0:
-			return nil
-		case 1:
-			return determineDispatchErr(res[2:])
+	var other Other
+	var invalidCustom InvalidCustom
+	var unknownCustom UnknownCustom
+
+	dispatchError := scale.MustNewVaryingDataType(other, CannotLookup{}, BadOrigin{}, Module{})
+	invalid := scale.MustNewVaryingDataType(Call{}, Payment{}, Future{}, Stale{}, BadProof{}, AncientBirthBlock{},
+		ExhaustsResources{}, invalidCustom, BadMandatory{}, MandatoryDispatch{})
+	unknown := scale.MustNewVaryingDataType(ValidityCannotLookup{}, NoUnsignedValidator{}, unknownCustom)
+
+	okRes := scale.NewResult(nil, dispatchError)
+	errRes := scale.NewResult(invalid, unknown)
+	result := scale.NewResult(okRes, errRes)
```

Comment:
I think these could be declared as package vars since they wouldn't ever change?

Reply:
Yup you are right, will do